packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+from typing import Protocol, runtime_checkable
+
+import pandas as pd
+from attrs import frozen
+from loguru import logger
+
+from ref_core.exceptions import ConstraintNotSatisfied
+
+
+@runtime_checkable
+class GroupValidator(Protocol):
+    """
+    A constraint that must be satisfied when executing a given metric run.
+
+    All constraints must be satisfied for a given group to be run.
```

Comment:
> group

I think it would be good to define the exact meaning of a group somewhere. I didn't find anything in the docs so far, and it seems the details are crucial here.